### DISCUSSION TITLE
Callback for each variable

### DISCUSSION THREAD

**Discussion by zedissime** [harttle/liquidjs#607] [Q&A]
Hello
Is there a way to call a callback to manipulate each variable ?

For example I've many templates who have some call of variable like those 
`{{ path_to_label }}` with data looking like this :
```
{
  path_to_label: {
      en: 'Label in english',
      es: 'Label in spanish'
  }
}
```
In the previous way to handle variables in strings, it was automatically taking the defined locale.

I was wondering if there was a way to handle this **without** using a filter. 
For example is there a callback to specify / override to tell how to get the value in the data object ?

**Comment by harttle** [comment]
It’s possible! Using Drop and method_missing will do the trick for you.

We will create docs for drops. Until then you can search for issues and test cases.

**Reply by zedissime** [reply]
Thank you for your prompt answer, but correct me if I'm wrong, the method_missing, is only able to do something when the var or path is missing.
In my case it is not, but `{{ path_to_label }}` returns `[object Object]`

**Reply by harttle** [reply]
Yes method_missing is only used when calling method/property. Thus you'll need an object first, top level variables cannot be returned by method_missing. Here's an example:

```javascript
const { Liquid, Drop } = require('liquidjs');

const mapping = {
   label: {
      en: 'Label in english',
      es: 'Label in spanish'
   }
}

class i18n extends Drop {
   constructor(lang) {
      super()
      this.lang = lang
   }
   liquidMethodMissing(name) {
      return mapping[name][this.lang]
   }
}

async function main() {
   const engine = new Liquid()
   const data = { strings: new i18n('en') }
   const template = `{{ strings.label }}`
   const html = await engine.parseAndRender(template, data)
   console.log(html)
}

main()
```

**Reply by harttle** [reply]
While using JavaScript getter, you can achieve reading top level variables. But the property name (`label` in this case) now need to be hardcoded.

```javascript
const { Liquid, Drop } = require('liquidjs');

const mapping = {
   label: {
      en: 'Label in english',
      es: 'Label in spanish'
   }
}

const lang = 'en'

const data = {
   get label() {
      return mapping.label[lang]
   }
}

async function main() {
   const engine = new Liquid()
   const template = `{{ label }}`
   const html = await engine.parseAndRender(template, data)
   console.log(html)
}

main()
```
